File: cpp/Fast and Slow Pointers/linked_list_midpoint.cpp
#include "ds/ListNode.h"
using ds::ListNode;

/**
 * Definition of ListNode:
 * struct ListNode {
 *     int val;
 *     ListNode* next;
 *     ListNode(int val = 0, ListNode* next = nullptr) : val(val), next(next) {}
 * };
 */

ListNode* linkedListMidpoint(ListNode* head) {
    ListNode* slow = head;
    ListNode* fast = head;
    // When the fast pointer reaches the end of the list, the slow
    // pointer will be at the midpoint of the linked list.
    while (fast != nullptr && fast->next != nullptr) {
        slow = slow->next;
        fast = fast->next->next;
    }
    return slow;
}

File: typescript/Fast and Slow Pointers/linked_list_midpoint.ts
import { ListNode } from "./ds";

/* Definition of ListNode:
class ListNode {
    val: number; next: ListNode | null;
    constructor(val: number) {
        this.val = val;
        this.next = null;
    }
}*/


function linkedListMidpoint(head: ListNode | null): ListNode | null {
    let slow: ListNode | null = head;
    let fast: ListNode | null = head;
    // When the fast pointer reaches the end of the list, the slow
    // pointer will be at the midpoint of the linked list.
    while (fast !== null && fast.next !== null) {
        slow = slow!.next;
        fast = fast.next.next!;
    }
    return slow;
}
